### PR TITLE
fix(acp): preserve JSON-RPC error data

### PR DESCRIPTION
**Pull request by brainx**
## Problem

`buzz-acp` currently keeps only a JSON-RPC error's `code` and `message`. Adapter diagnostics carried in `error.data` are lost, turning actionable failures such as an incompatible CLI flag into a generic `-32603 Internal error`.

## Fix

Preserve `error.data` as a compact, UTF-8-safe, 4 KiB-bounded owner diagnostic. High-level logs and channel failure notices continue to use only the public code and message; the encrypted `turn_error` observer payload receives the detail in a separate `data` field.

Errors without a string `message` retain their bounded full error object for owner diagnosis. The implementation retains immediate handling for non-retryable authentication failures.

## Regression guard

Add focused tests for structured data, missing and null messages, UTF-8 truncation, public display behavior, observer propagation, and owner-transcript rendering.

## Testing

- `cargo fmt -p buzz-acp -- --check` — passed
- `cargo test -p buzz-acp` — 673 unit tests and 9 lifecycle tests passed
- `node --import ./test-loader.mjs --experimental-strip-types --test src/features/agents/ui/agentSessionTranscript.test.mjs` — 53 passed
- `pnpm exec biome check src/features/agents/lib/friendlyAgentLastError.ts src/features/agents/ui/agentSessionTranscript.ts src/features/agents/ui/agentSessionTranscript.test.mjs` — passed
- `git diff --check upstream/main...HEAD` — passed

Addresses the diagnostic-loss portion of #2422.

### CODE REVIEW

**Comment by SeanGearin** [comment]
Solid root-cause fix — preserving `data` structurally is clearly right for tooling that wants to parse it. One thing worth flagging on the desktop consumer side, since it affects whether operators actually *see* the detail you're now preserving: the transcript builds its user-facing text from `payload.error` / `payload.code` (`agentSessionTranscript.ts:775-776`) and never reads `payload.data`, and `friendlyAgentLastError`'s `-32603` branch (written to preserve adapter detail after the standard prefix) never receives any, because buzz-acp strips it upstream. So today the detail lands in the observer payload but not in front of the operator.

Relatedly, the codex reviewer's P2 on the `"Unknown agent error"` message-missing fallback looks real: for data-only errors that string becomes all a string-based consumer gets, where `error.to_string()` previously at least carried the raw JSON. If you also appended the bounded `data` onto the `AgentError` message itself (in addition to the structured field), the existing `Display`/`to_string()` surfaces — harness logs, the channel retry notice, and that `-32603` preserve-detail branch — would light up with no new plumbing, and the fallback wouldn't drop detail. Structured field for tooling + message append for humans compose cleanly. Happy to help with the message-side piece if useful.

File: crates/buzz-acp/src/acp.rs [hunk 1]
```diff
@@ -105,20 +108,48 @@ pub enum AcpError {
...
+/// preserving the numeric code and bounded provider-specific `data`. When the
+/// `message` field is missing or non-string, use a stable fallback message while
+/// retaining any `data` detail.
 fn agent_error_from_json(error: &serde_json::Value) -> AcpError {
     let code = error.get("code").and_then(|c| c.as_i64()).unwrap_or(-32000);
-    let message = match error.get("message").and_then(|m| m.as_str()) {
-        Some(m) => m.to_string(),
-        None => error.to_string(),
+    let wire_message = error.get("message").and_then(|message| message.as_str());
+    let message = wire_message.unwrap_or("Unknown agent error").to_string();
```
[the remark this answers is not in the document]
Comment (brainx) [reply crates/buzz-acp/src/acp.rs:125]: Addressed in d835077c. The desktop owner transcript now consumes the bounded `turn_error.data` field, so data-only errors and `-32603` adapter diagnostics remain visible instead of collapsing to `Unknown agent error`. I kept `AcpError`'s public `Display` unchanged so raw diagnostics do not spread into general logs or channel retry notices.

Regression coverage is in `agentSessionTranscript.test.mjs`; the full `just ci` gate passes.

**Comment by brainx** [comment]
@SeanGearin Addressed in `d835077c` by consuming the bounded `turn_error.data` field in the desktop owner transcript. I kept it out of `AcpError::Display` because that text also feeds general logs and channel retry notices. This keeps the diagnostic human-visible to the operator without broadening its exposure. Regression coverage includes both data-only and `-32603` errors.

**Comment by brainx** [comment]
Updated the branch to current `main` in `f7ef14d3`. The conflict resolution preserves both the bounded owner-only `error.data` diagnostics and the newer immediate handling for non-retryable authentication errors. Current-head verification passed: `cargo test -p buzz-acp` (603 unit + 9 lifecycle tests), the 51 transcript tests, Rust formatting, and targeted Biome checks. DCO is green; the refreshed CI and Docker workflow runs are awaiting maintainer approval.

**Comment by SeanGearin** [comment]
Verified d835077c against the head - `friendlyTurnErrorCopy` consuming `payload.data` with the transcript-only scoping is exactly the shape I was hoping for, and pinning both the data-only `-32000` and `-32603` cases means the visibility can't silently regress. Keeping it out of `Display` was the right call for the log/notice surfaces. Nothing further from me - this looks ready for a maintainer pass.

**Comment by munirad7s** [comment]
Security note before preserving error.data: serializing the full adapter-provided data object into logs can newly expose credentials that were previously hidden by this bug. Please recursively redact credential-shaped keys (for example authorization, api_key, token, secret, password) and apply a separate bounded length cap to the rendered data. We have a tested one-file fork implementation covering nested redaction and truncation and can provide the patch if useful.

**Comment by brainx** [comment]
@munirad7s Addressed in 2677d42f.

Adapter-provided error data is now recursively redacted before serialization, including nested objects and arrays. Values under credential-shaped keys such as authorization, api_key, token, secret, password, cookie, private key, and bearer are replaced with [REDACTED], while safe diagnostic fields remain intact. The separate existing 4 KiB UTF-8-safe rendered-data cap remains in place.

Regression coverage includes mixed key styles, nested credentials, whole credential objects, and preservation of safe fields. Full local verification: just ci passed.

**Comment by brainx** [comment]
@munirad7s Follow-up hardening is now in 7899985b0.

After my earlier reply, I found one additional ACP-specific serialization shape: session/new MCP environment values are carried as name/value records, so matching only sensitive JSON field names was insufficient. This follow-up recognizes sensitive identifiers in name/value and key/value records and redacts their value, covering the actual BUZZ_PRIVATE_KEY and BUZZ_AUTH_TAG wire shape as well as direct auth_tag fields. Non-secret entries such as BUZZ_RELAY_URL remain visible, and the UTF-8-safe 4 KiB bound is unchanged.

Regression coverage uses the exact session/new.params.mcpServers[].env[] structure.

Verification:
- cargo test -p buzz-acp — 782 unit tests and 9 lifecycle tests passed
- cargo clippy -p buzz-acp --all-targets -- -D warnings — passed
- CI=true just ci — passed
